Data/Dimensions/SI.hs

```diff
@@ -26,6 +26,19 @@ declareDimension "Temperature"
 declareDimension "AmountOfSubstance"
 declareDimension "LuminousIntensity"
 
+-- | A plane angle is dimensionless; alternatively, it has dimension
+-- length/length. It would be wrong to divide 2 meters by 1
+-- meter and conclude that the quantity is 2 radians or degrees. To make
+-- plane angle safe to use, we define it as a fundamental dimension.
+declareDimension "PlaneAngle"
```

Comment:
I feel like I need more convincing that this should be a new dimension. As you've defined it, this dimension is unrelated to the dimension of length/length -- you would need some explicit conversion factor of 1.

You're right that not having this as a dimension could allow trouble, but there are many ways for trouble to arise.

I'm happy to take the new units (degrees, etc.), but I'm less convinced on these new dimensions.

Reply:
> this dimension is unrelated to the dimension of length/length

That's exactly what I want as a user of this library who is trying to use the type system to prevent logical errors. Maybe this whole discussion would have been avoided if the highlighted code comment didn't exist?

> this dimension is unrelated to the dimension of length/length -- you would need some explicit conversion factor of 1

There is no single explicit conversion factor that relates a ratio of lengths to an angle (i.e. the relationship is nonlinear and function-dependent). Thus, there should be no implicit way of converting from a ratio of lengths (a dimensionless quantity) to an angle.

Consider the arcsine function: asin(1 meter/2 meter) = 30 degrees, so the conversion factor there is 60 (to go from a dimensionless 1/2 to 30 degrees). Meanwhile, asin(1 meter/sqrt(2) meters) = 45 degrees, so the conversion factor there is approximately 63.6. There is no single conversion factor, for the arcsine function, from a ratio of lengths to an angle.

Furthermore, there are many functions which relate a ratio of lengths to an angle: arccosine, arctangent, inverse hyperbolic tangent, all the inverses of the preceding functions (e.g. cosine), etc. So even if there was a single conversion factor for arcsine, it would still be wrong to model an angle as a dimensionless quantity, as the angle vs ratio of lengths relationship might be via one of these other functions. 

In summary, we can't convert an arbitrary ratio of lengths to an angle. We must know how the two lengths relate to each other (and sometimes they don't; you can't just apply sine/arcsine to an abitrary pair of divided lengths) to correctly go from ratio to angle, or angle to ratio.

Finally, if we modeled an angle as a dimensionless quantity, then we could erroneously convert a ratio of seconds to an angle, or a ratio of forces, etc.

> You're right that not having this as a dimension could allow trouble, but there are many ways for trouble to arise.

I don't see how this is an argument against making angles a new kind of dimension. You could apply your same argument to this library as a whole (there are many ways for software to be wrong, besides making errors in dimensions), or more generally to type safety (Haskell's type system doesn't prevent every kind of error), but we still prefer to make certain classes of errors impossible, via the type system. I'm sure you understand all of this quite well, so perhaps I misunderstood your meaning...?

I think it boils down to "let's do what's useful for the library's users". Making angle a new kind of dimension does not prevent us from doing correct things:

* I can still add two angles together to make a new angle.
* I can divide an angle by a dimensionless constant to make a new angle.
* I can convert between radians and degrees.
* I can take the opposite side and hypoteneuse of a right triangle, divide them, extract the Double, pass that to arcsine, then make it a quantity of Radians. In fact, my application which uses this forked library happily does this, and much more with angles, and I'm very pleased with PlaneAngle as a dimension.

Making angle a new kind of dimension does prevent errors:

* I can't do some_length/another_length and then use it in the place of an angle.
* I can't do some_time/another_time and then use it in the place of an angle.
* I can't do some_angle/another_angle and then use it in the place of an angle.

It seems to me that this library was developed precisely to eliminate the errors described above while allowing the correct things described above, for other dimensioned quantities. Why not for angles?

Finally, in #10 you said "I'm happy to have Angle and such in units-defs", so I wonder why you've changed your mind...? nushio3 and jameysharp agreed that angles should be a kind of dimension.

Reply:
You make a convincing argument. I think where I stumble is in e.g. the definition of `arcsin`. In your examples, it would take us from a dimensionless quantity to an Angle. That's about right. But this commit doesn't include that facility. Maybe it should? Or maybe the theory of `units-defs` is just to provide the ability for some other library to do this?

Reply:
I like your suggestion of providing a unit-aware arcsin, arccos, sin, cos, etc., but I feel those are outside the scope of units-defs. So I lean towards "the theory of units-defs is just to provide the ability for some other library to do this".

I think the design space admits several possibilities, and so I think it's not right for units-defs to choose one. Better to let the application decide what it needs (by including one of several options for this facility).
For example, there are a number of dimensions where you could vary the design of `arcsin` and friends:

* `arcsin` should work for both Float and Double. So, naturally we'd place a constraint on `arcsin` so that both arguments are Floating and Length. However, this doesn't play nicely if someone wants to use a different numerical typeclass hierarchy (i.e. does not use `Floating`). 

* Naturally, we'd reuse the `base` package's implementation of `asin`. What if someone wants to use a different implementation of `asin` (e.g. faster but more approximate)? 

Since there is no single obvious correct answer to the questions above, I'd rather delegate the design and implementation to another library (or another, later commit to this library).

Finally, I don't have such code written today (and may never write such code). Let's not prevent useful code from entering the library while waiting for code that may never materialize.

Reply:
OK. You win. :)

Reply:
We all win! Cheers, mate  :D